### PR TITLE
Add LocalTopicModel object

### PR DESCRIPTION
**Pull request by charleslparker**
@mmerce - Here's the LocalTopicModel object that more or less matches the equivalent in the python bindings.  I have duplicated the test we have there, but do not yet have tests against the server as I believe we discussed earlier.

I've also not yet added appropriate doc to `docs`, but it looks like I could do that now if this is a good time, or in a seperate PR after everything is merged (which I'd probably prefer so you can give it a thorough check).

Obviously I've tried to follow your conventions but let me know if I've missed anything crucial.  I've also tried to keep variable names as similar to the python bindings as possible.

I believe @felmarlop needs this for his work ASAP.  Felix, you can test against my fork until this is merged if that's possible.

Finally, do let me know if you'd like the PR against a different branch/repo.

### CODE REVIEW

**Comment by mmerce** [comment]
@charleslparker and @felmarlop - I saw today that @leonhwang has a version which might be used to test against when we check it works, but I'd probably test first the backend, just to make sure we agree that the remote JSON is what we expect. Only then would I think about the local prediction part. Do you agree @felmarlop?

**Comment by felmarlop** [comment]
Yes, I agree. Im having problems with my branch with LDA related to the new names and more things. When  I have solved that I'll check the remote JSON. 
Anyway, I'll work on lda local predictions after next Friday and right now I have to adapt the new name and display the UI with LDA only using the remote response, no local predictions yet.

**Comment by charleslparker** [comment]
@mmerce @felmarlop - Sounds good to me.

**Comment by felmarlop** [comment]
The Leon's apian with LDA is ready to merge in next. About these local predictions we would need all the vocabulary in order to calculate the probabilities, isn't it?, I'd need some instructions to use them, I think they are a little different from the rest.

**Comment by charleslparker** [comment]
>  About these local predictions we would need all the vocabulary in order to calculate the probabilities, isn't it?

On this branch of @mmerce's, the vocabulary has already been changed, I think.

> I'd need some instructions to use them, I think they are a little different from the rest.

```
var localModel = bigml.LocalTopicModel("topicmodel/id")
var prediction = localModel.distribution({"text field": "Some text in the text field"})
```

`prediction` is an array of float-point numbers - the topic distribution of the input data.  As far as I can tell, this roughly the same as `LocalCluster`.  Do let me know if it isn't.

**Comment by felmarlop** [comment]
Thanks Charly! I'll let you know if I have problems.

File: index.js [hunk 1]
```diff
@@ -76,5 +76,7 @@ module.exports = {
   // Local Logistic Regression object for local predictions
   LocalLogisticRegression: require('./lib/LocalLogisticRegression'),
   // Local Logistic Regression object for local predictions
```
Comment (mmerce) [inline index.js:82]: comment should be changed to the correct local resource name. Probably done if you previously rebase to my topic branch

File: index.js [hunk 2]
```diff
@@ -76,5 +76,7 @@ module.exports = {
...
-  LocalAssociation: require('./lib/LocalAssociation')
+  LocalAssociation: require('./lib/LocalAssociation'),
+  // Local Logistic Regression object for local predictions
```
Comment (mmerce) [inline index.js:84]: like the previous one

File: lib/LocalTopicModel.js [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+    "es": "spanish",
+    "sv": "swedish",
+    "tr": "turkish"
+}
+
+if (NODEJS) {
+  var util = require('util');
+  var events = require('events');
+  var TopicModel = require(PATH + 'TopicModel');
+  var Snowball = require('snowball');
```
Comment (mmerce) [inline lib/LocalTopicModel.js:56]: the modules included here are only required if executed from node.js. The javascript version will not use them and they are usually related to functions not used in the local Topic Model code, but to the retrieval of the JSON resource. Are the snowball and mersenne-twister libraries correctly here?

File: lib/LocalTopicModel.js [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+      throw new Error('Cannot create the TopicModel instance. Could not' +
+                      ' retrieve the resource: ' + error);
+
+    }
+
+    status = utils.getStatus(resource);
+    if ((typeof resource.object) !== 'undefined') {
+      resource = resource.object;
+    }
+    if ((typeof resource.model) !== 'undefined') {
```
Comment (mmerce) [inline lib/LocalTopicModel.js]: I believe now the info is in the `topic_model` attribute. Please check.

File: lib/LocalTopicModel.js [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+        }
+
+        self.termToIndex = {};
+
+        for (i = 0; i < model.termset.length; i++) {
+          tstem = self.stem(model.termset[i]);
+          self.termToIndex[tstem] = i;
+        }
+
+        self.seed = model.hashed_seed;
```
Comment (mmerce) [inline lib/LocalTopicModel.js]: I only see `seed` and `topicmodel_seed` as attributes in the model info

File: lib/LocalTopicModel.js [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+        self.caseSensitive = model.case_sensitive;
+        self.bigrams = model.bigrams;
+
+        self.ntopics = model.term_topic_assignments[0].length;
```
Comment (mmerce) [inline lib/LocalTopicModel.js:161]: I don't see `term_topic_assignments` as attribute

File: lib/LocalTopicModel.js [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+        self.alpha = model.alpha;
```
Comment (mmerce) [inline lib/LocalTopicModel.js:162]: I don't see `alpha` as attribute

File: lib/LocalTopicModel.js [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+          self.uniformDocAssignments[i] = 1;
+        }
+
+        self.uniformNormalizer = self.ntopics + self.ktimesalpha;
+
+        self.temp = new Float64Array(self.ntopics);
+
+        assignments = model.term_topic_assignments;
+        beta = model.beta;
+        nterms = model.termset.length;
```
Comment (mmerce) [inline lib/LocalTopicModel.js:177]: Nor do I see beta or termset. I see:

topic_model: {
bigrams: false,
case_sensitive: false,
fields: {...},
language: "en",
number_of_topics: 12,
term_limit: 4096,
top_n_terms: 10,
topicmodel_seed: "BigML",
topics: [...]
},

File: lib/LocalTopicModel.js [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+  var char, index, lastTerm, termBefore, buf, tstem,
+      spaceWasSep, sawChar, outTerms;
+
+  buf = "";
+  outTerms = [];
+  sawChar = spaceWasSep = false;
+  lastTerm = termBefore = null;
+
+  index = 0;
+
```
Comment (mmerce) [inline lib/LocalTopicModel.js:264]: the previous inital assignments can be done in the previous declaration line

**Review by mmerce — COMMENTED** [review]
I might be wrong, but I cannot find many JSON attributes in the resource in our development instance. Please check that. And I'll have a second look as to style.

File: lib/LocalTopicModel.js (hunk 4 above)
Comment (felmarlop) [reply lib/LocalTopicModel.js]: Yes, it's topic_model now. It's the first error that the module gave me and I was gonna report it right now.

**Comment by mmerce** [comment]
@charleslparker, I will cherry-pick your last two commits and merge them to my current local-topic branch. I believe the rest of commits are already there, so I will close this PR. Does it sound good?

**Comment by mmerce** [comment]
Closing PR. Changes merged in local-topic branch.

**Comment by charleslparker** [comment]
I just pushed one more commit in https://github.com/mmerce/bigml-node/pull/5.  Thanks, @mmerce !  @felmarlop : this should fix the issues you're seeing.

**Pull request closed without merging**